modules/nlu/src/backend/tools/strings.ts

```diff
@@ -0,0 +1,78 @@
+import _ from 'lodash'
+
+/** Splits a string in all its N-grams */
+export function ngram(value: string, n: number): string[] {
+  const nGrams: string[] = []
+  let index = value.length - n + 1
+
+  while (index--) {
+    nGrams.push(value.slice(index, index + n))
+  }
+
+  return nGrams
+}
+
+export function vocabNGram(tokens: string[]): string[] {
+  const plainTokens = tokens
+    .map(x => x.replace('_', '\u2581')) // We want to discover real language-specific chars
+    .filter(x => x.length > 1) // We want to exclude tokens that represent ponctuation etc (tokenizers will often split them alone)
+
+  // We build a gramset, which is essentially a list of all the unique bigrams and trigrams
+  // We'll create entirely new words from those grams
+  const gramset = _.chain(plainTokens)
+    .map(x => [ngram(x, 1), ngram(x, 2)])
+    .flattenDeep()
+    .uniq()
+    .value()
+
+  return (gramset as never) as string[]
+}
+
+/** Returns the similarity of two sets of strings in percentage */
+export function setSimilarity(a: string[], b: string[]): number {
+  const common = _.intersection(a, b).length
+  return common / (a.length + b.length - common)
+}
+
+/**
+ * Returns the levenshtein distance between two strings
+ * @returns the # of operations required to go from a to b
+ */
+export function levenshtein(a: string, b: string): number {
```

Comment:
if I recall we already have this in toolkit  ?

Reply:
yes but quite innacessible from NLU and didn't want to involve dependency injection for such a simple function.  I'll add a comment to remove that when/if we refactor NLU in core